test/test_tracer.cpp

```diff
@@ -1175,3 +1176,116 @@ TEST_CASE(
   REQUIRE(found->second ==
           test_case.expected_error_prefix + test_case.tid_tag_value);
 }
+
+TEST_CASE("heterogeneous extraction") {
+  // These test cases verify that when W3C is among the configured extraction
+  // styles, then non-Datadog and unexpected Datadog fields in an incoming
+  // `tracestate` are extracted, under certain conditions, even when trace
+  // context was extracted in a non-W3C style.
+  //
+  // The idea is that a tracer might be configured to extract, e.g.,
+  // [DATADOG, B3, W3C] and to inject [DATADOG, W3C]. We want to make
+  // sure that no W3C-relevant information from the incoming request is lost in
+  // the outgoing W3C headers, even if trace context is extracted on account of
+  // DATADOG or B3.
+  //
+  // See the `TestCase` instances, below, for more information.
+
+  class MockIDGenerator : public IDGenerator {
+   public:
+    TraceID trace_id(const TimePoint&) const override {
+      throw std::logic_error("This test should not generate a trace ID.");
+    }
+    std::uint64_t span_id() const override { return 0x2a; }
+  };
+
+  struct TestCase {
+    int line;
+    std::string description;
+    std::vector<PropagationStyle> extraction_styles;
+    std::vector<PropagationStyle> injection_styles;
+    std::unordered_map<std::string, std::string> extracted_headers;
+    std::unordered_map<std::string, std::string> expected_injected_headers;
+  };
+
+  // clang-format off
+  auto test_case = GENERATE(values<TestCase>({
+    {__LINE__, "tracestate from primary style",
+     {PropagationStyle::W3C, PropagationStyle::DATADOG},
+     {PropagationStyle::W3C},
+     {{"traceparent", "00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01"},
+      {"tracestate", "dd=foo:bar,lol=wut"}},
+     {{"traceparent", "00-4bf92f3577b34da6a3ce929d0e0e4736-000000000000002a-01"},
```

Comment:
The parent ID that was extracted is the ID of some remote span (in the calling service). Then we create a new span, with its own ID, and inject its information. So, the injected span ID is the ID that we generated for our one local span.

It's `0x2a` because of `MockIDGenerator`.